skorch/tests/callbacks/test_lr_scheduler.py

```diff
@@ -94,10 +94,10 @@ def test_lr_callback_steps_correctly(
         )
         net.fit(X, y)
         # pylint: disable=protected-access
-        assert lr_policy.lr_scheduler_.last_epoch == max_epochs - 1
+        assert lr_policy.lr_scheduler_.last_epoch == max_epochs
 
     @pytest.mark.parametrize('policy, kwargs', [
-        (TorchCyclicLR, {'base_lr': 1e-3, 'max_lr': 6e-3}),
+        (TorchCyclicLR, {'base_lr': 1e-3, 'max_lr': 6e-3, 'step_every': 'batch'}),
```

Comment:
This would not be backwards compatible with `TorchCyclicLR` with default arguments. As in `TorchCyclicLR` would step every epoch if a user does not pass in `step_every='batch'`.

Reply:
So the implementation is fine and we just need to restore the ` if TorchCyclicLR and isinstance(self.lr_scheduler_, TorchCyclicLR):...` segment?

Reply:
Now that `step_every` is a parameter to `LRScheduler`, if we keep the `isinstance(self.lr_scheduler_, TorchCyclicLR)` piece, we will not be strictly following the `step_every` parameter.

We can have a `step_every='auto'` option that special cases the `CosineAnnealingLR` and `TorchCyclicLR`. Specifically, if the class is `CosineAnnealingLR` or `TorchCyclicLR` we use batch steps, otherwise we use epoch step. We will still allow the `batch` and `epoch` options for custom lr schdulers.

Reply:
That would mean introducing the entire concept of an 'auto' option ,and having to be backwards compatible for it in the future. However all it'll do is special case cyclic (for the record, cosine annealing isn't hard coded to work with batches, only cyclic). 

This basically means that if we don't want 'auto'  to just be a cover up for the cyclic case and instead be meaningful, we will have to add reasonable recommendations for every scheduler we introduce (which will be the default auto options). That will most likely involve having to read the paper from which it was introduced, or it's documentation.

Reply:
This is a bit of a conundrum. Do you know why some schedulers prefer batch and other epoch? Is it some derived property or more arbitrary?

> we will have to add reasonable recommendations for every scheduler we introduce

This is certainly not the way to go. Could we not just use whatever the default is?

Reply:
> This is a bit of a conundrum. Do you know why some schedulers prefer batch and other epoch? Is it some derived property or more arbitrary?

Not for certain. Naturally if you wanted your lr to change smoothly you'd do it per batch rather than epoch. I'm just plugging in a scheduler as is written in its documentation.

> This is certainly not the way to go. Could we not just use whatever the default is?

A scheduler doesn't have a default. It has a step function which the user can use whenever and wherever they'd like. Setting mode `epoch` or `batch` as defaults will break backwards compatibility for anything that isn't intended to use the default mode we choose.

Since the Cyclic scheduler is a special case for which skorch hacked a solution, I suggested leaving that solution as is (so that it won't conflict with this feature or have compatibility issues) until skorch makes an update that breaks compatibility, and just get rid of that hack then.

Reply:
Good points. Maybe it would be worth it to actually break backwards compatibility here for the sake of getting a clean solution. In fact, we could even detect when such a breakage happens and issue a warning to the user, with instructions how to revert to the old behavior. Would that be a compromise?

Reply:
> Would that be a compromise?

I would be happy with a deprecation warning with a suggestion of using `step_every='batch'` in the future.

Reply:
Done.